### ISSUE TITLE
Segfault when using `module.create_jit_execution_engine()`

### ISSUE THREAD

**Issue by crawdaddie** [TheDan64/inkwell#452]
**Describe the Bug**
Segfault when using `module.create_jit_execution_engine()`
the jit can be used for compiling and executing the function correctly, but by the time the execution engine is dropped a segfault occurs

**To Reproduce**
adapted from the inkwell kaleidoscope example
```rust
use inkwell::builder::Builder;
use inkwell::context::Context;
use inkwell::execution_engine::{ExecutionEngine, JitFunction};
use inkwell::module::Module;
use inkwell::OptimizationLevel;

use std::error::Error;

/// Convenience type alias for the `sum` function.
///
/// Calling this is innately `unsafe` because there's no guarantee it doesn't
/// do `unsafe` operations internally.
type SumFunc = unsafe extern "C" fn(u64, u64, u64) -> u64;

struct CodeGen<'ctx> {
    context: &'ctx Context,
    module: Module<'ctx>,
    builder: Builder<'ctx>,
    execution_engine: ExecutionEngine<'ctx>,
}

impl<'ctx> CodeGen<'ctx> {
    fn jit_compile_sum(&self) -> Option<JitFunction<SumFunc>> {
        let i64_type = self.context.i64_type();
        let fn_type = i64_type.fn_type(&[i64_type.into(), i64_type.into(), i64_type.into()], false);
        let function = self.module.add_function("sum", fn_type, None);
        let basic_block = self.context.append_basic_block(function, "entry");

        self.builder.position_at_end(basic_block);

        let x = function.get_nth_param(0)?.into_int_value();
        let y = function.get_nth_param(1)?.into_int_value();
        let z = function.get_nth_param(2)?.into_int_value();

        let sum = self.builder.build_int_add(x, y, "sum");
        let sum = self.builder.build_int_add(sum, z, "sum");

        self.builder.build_return(Some(&sum));

        unsafe { self.execution_engine.get_function("sum").ok() }
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    let context = Context::create();
    let module = context.create_module("sum");
    let execution_engine = module.create_jit_execution_engine(OptimizationLevel::None)?;

    let codegen = CodeGen {
        context: &context,
        module,
        builder: context.create_builder(),
        execution_engine,
    };

    let sum = codegen
        .jit_compile_sum()
        .ok_or("Unable to JIT compile `sum`")?;

    let x = 1u64;
    let y = 2u64;
    let z = 3u64;

    unsafe {
        println!("{} + {} + {} = {}", x, y, z, sum.call(x, y, z));
        assert_eq!(sum.call(x, y, z), x + y + z);
    }

    Ok(())
}

```

**Expected Behavior**
the program should not segfault after executing the `sum` function

**LLVM Version (please complete the following information):**
 - LLVM Version: 16.0
 - Inkwell Branch Used: llvm16-0

**Desktop (please complete the following information):**
- OS: [e.g. Ubuntu 16.04, Windows 10] MacOS

**Comment by JCBurnside** [comment]
from what I can tell it's invalid to change a module after it has been added to an execution engine.  My recommendation would be to make `Module::create_jit_exectuion_engine` consume self.

**Comment by crawdaddie** [comment]
closing this as it appears to be fixed with inkwell 0.4.0

**Comment by crawdaddie** [comment]
closing this as it appears to be fixed with inkwell 0.4.0
